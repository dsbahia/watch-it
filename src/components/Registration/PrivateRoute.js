import { Route, Navigate } from "react-router-dom";
import { useAuthValue } from "./AuthContext";

function PrivateRoute({ component: Component, ...rest }) {
  const { currentUser } = useAuthValue();

  return (
    <Route
      {...rest}
      render={(props) => {
        return currentUser.emailVerified ? (
          <Component {...props} />
        ) : (
          <Navigate to="/login" />
        );
      }}
    />
  );
}
export default PrivateRoute;
