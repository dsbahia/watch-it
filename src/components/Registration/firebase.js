import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";

const firebaseConfig = {
  apiKey: "AIzaSyDpiscXs7UomLcbyhm4Lq8rdp6zUoGiy-s",
  authDomain: "watch-it-be.firebaseapp.com",
  projectId: "watch-it-be",
  storageBucket: "watch-it-be.appspot.com",
  messagingSenderId: "241617376848",
  appId: "1:241617376848:web:5c5ec7579fc6567bed6355",
  measurementId: "G-841E4LK74N",
};
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
export { auth };
