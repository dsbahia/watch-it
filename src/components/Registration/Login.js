import { React, useState } from "react";
import {
  signInWithEmailAndPassword,
  sendEmailVerification,
} from "firebase/auth";
import { Link, useNavigate } from "react-router-dom";
import { auth } from "./firebase";
import { useAuthValue } from "./AuthContext";
import "../../styles/forms.css";

function Login() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const { setTimeActive } = useAuthValue();
  const navigate = useNavigate();

  const login = (e) => {
    e.preventDefault();
    signInWithEmailAndPassword(auth, email, password)
      .then(() => {
        if (!auth.currentUser.emailVerified) {
          sendEmailVerification(auth.currentUser)
            .then(() => {
              setTimeActive(true);
              navigate("/verify-email");
            })
            .catch((err) => alert(err.message));
        } else {
          navigate("/");
        }
      })
      .catch((err) => setError(err.message));
  };

  return (
    <div className="auth-container">
      <div className="auth-form">
        <h1>Log in</h1>
        {error && <div className="auth-error">{error}</div>}
        <form onSubmit={login} name="login_form">
          <input
            type="email"
            value={email}
            required
            placeholder="Enter your email"
            onChange={(e) => setEmail(e.target.value)}
            className="auth-input"
          />

          <input
            type="password"
            value={password}
            required
            placeholder="Enter your password"
            onChange={(e) => setPassword(e.target.value)}
            className="auth-input"
          />

          <button type="submit" className="auth-button">
            Login
          </button>
        </form>
        <span>
          Don&apos;t have an account?
          <Link to="/register">Create one here</Link>
        </span>
      </div>
    </div>
  );
}

export default Login;
