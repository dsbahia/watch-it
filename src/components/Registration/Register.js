import { useState } from "react";
import { useNavigate, Link } from "react-router-dom";
import {
  createUserWithEmailAndPassword,
  sendEmailVerification,
} from "firebase/auth";
import { auth } from "./firebase";
import { useAuthValue } from "./AuthContext";
import "../../styles/forms.css";

function Register() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [error, setError] = useState("");
  const { setTimeActive } = useAuthValue();

  const navigate = useNavigate();

  const validatePassword = () => {
    let isValid = true;
    if (password !== "" && confirmPassword !== "") {
      if (password !== confirmPassword) {
        isValid = false;
        setError("Passwords do not match");
      }
    }
    return isValid;
  };

  const register = (e) => {
    e.preventDefault();
    setError("");
    if (validatePassword()) {
      createUserWithEmailAndPassword(auth, email, password)
        .then(() => {
          sendEmailVerification(auth.currentUser);
        })
        .then(() => {
          setTimeActive(true);
          navigate("/verify-email");
        })
        .catch((err) => alert(err.message));
    }
    setEmail("");
    setPassword("");
    setConfirmPassword("");
  };

  return (
    <div className="center">
      <div className="auth">
        <h1>Register</h1>
        {error && <div className="auth__error">{error}</div>}
        <form onSubmit={register} name="registration_form">
          <input
            type="email"
            value={email}
            placeholder="Enter your email"
            required
            onChange={(e) => setEmail(e.target.value)}
          />

          <input
            type="password"
            value={password}
            required
            placeholder="Enter your password"
            onChange={(e) => setPassword(e.target.value)}
          />

          <input
            type="password"
            value={confirmPassword}
            required
            placeholder="Confirm password"
            onChange={(e) => setConfirmPassword(e.target.value)}
          />

          <button type="submit">Register</button>
        </form>
        <span>
          Already have an account?
          <Link to="/login">login</Link>
        </span>
      </div>
    </div>
  );
}

export default Register;
