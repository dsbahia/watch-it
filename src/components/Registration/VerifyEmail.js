import { useState, useEffect } from "react";
import { sendEmailVerification } from "firebase/auth";
import { auth } from "./firebase";
import { useNavigate } from "react-router-dom";
import { useAuthValue } from "./AuthContext";
import "../../styles/verifyEmail.css";

function VerifyEmail() {
  const { currentUser } = useAuthValue();
  const [time, setTime] = useState([]);
  const [endTime, setEndTime] = useState([]);
  const { timeActive, setTimeActive } = useAuthValue();
  const navigate = useNavigate();

  const startWork = (e) => {
    e.preventDefault();
    const StartTime = {
      thisTime: new Date().toLocaleString(),
    };
    setTime(StartTime);
  };

  const finishWork = (e) => {
    e.preventDefault();
    const StopTime = {
      endTime: new Date().toLocaleString(),
    };
    setEndTime(StopTime);
  };
  const resendEmailVerification = () => {
    sendEmailVerification(auth.currentUser)
      .then(() => {
        setTimeActive(true);
      })
      .catch((err) => {
        alert(err.message);
      });
  };
  return (
    <div className="center">
      <div className="verify-email">
        <h1>Verify your Email Address</h1>
        <p>
          <strong>A Verification email has been sent to:</strong>
          <br />
          <span>{currentUser?.email}</span>
        </p>
        <span>Follow the instruction in the email to verify your account</span>
        <button
          type="button"
          onClick={resendEmailVerification}
          disabled={endTime.endTime}
        >
          Resend Email
          {timeActive && time}
        </button>
      </div>
    </div>
  );
}

export default VerifyEmail;
