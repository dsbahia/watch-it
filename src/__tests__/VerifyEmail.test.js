import React from "react";
import { render, fireEvent, screen, act } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { toast } from "react-hot-toast";
import { useAuthValue } from "../components/Registration/AuthContext";
import VerifyEmail from "../components/Registration/VerifyEmail";

jest.mock("react-hot-toast", () => ({
  error: jest.fn(),
}));

jest.mock("../components/Registration/AuthContext", () => ({
  useAuthValue: jest.fn(),
}));

beforeEach(() => {
  useAuthValue.mockReturnValue({
    currentUser: null,
  });
});

afterEach(() => {
  jest.clearAllMocks();
});

describe("VerifyEmail", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <VerifyEmail />
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display a button and trigger an email to the user on click", () => {
    render(
      <MemoryRouter>
        <VerifyEmail />
      </MemoryRouter>,
    );
    const verifyButton = screen.getByText("Resend Email");

    fireEvent.click(verifyButton);
  });

  it("should render a message if an email has been sent", () => {
    const isEmailSent = true;
    render(
      <MemoryRouter>
        <VerifyEmail isEmailSent={isEmailSent} />
      </MemoryRouter>,
    );
    if (isEmailSent) {
      const emailSentMessage = screen.getByText(
        "A Verification email has been sent to:",
      );
      expect(emailSentMessage).toBeInTheDocument();
    } else {
      expect(toast.error).toHaveBeenCalledWith(
        "An error occurred. Please try again later.",
      );
    }
  });

  it('starts and stops the countdown timer when "Resend Email" is clicked', async () => {
    const { getByText, findByText } = render(
      <MemoryRouter>
        <VerifyEmail />
      </MemoryRouter>,
    );

    const resendButton = getByText("Resend Email");

    expect(resendButton).not.toHaveAttribute("disabled");

    await act(async () => {
      await findByText("Resend Email");
    });

    expect(resendButton).not.toHaveAttribute("disabled");
  });
});
