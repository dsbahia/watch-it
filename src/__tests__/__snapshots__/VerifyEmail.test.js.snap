// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyEmail matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="center"
  >
    <div
      class="verify-email"
    >
      <h1>
        Verify your Email Address
      </h1>
      <p>
        <strong>
          A Verification email has been sent to:
        </strong>
        <br />
        <span />
      </p>
      <span>
        Follow the instruction in the email to verify your account
      </span>
      <button
        type="button"
      >
        Resend Email
      </button>
    </div>
  </div>
</DocumentFragment>
`;
